shell/browser/api/electron_api_utility_process.cc

```diff
@@ -0,0 +1,642 @@
+                       utility_process_wrapper_,
+                       api::UtilityProcessWrapper::IOHandle::STDERR,
+                       std::move(message)));
+  }
+
+  void ShutdownReader() override {
+    content::GetUIThreadTaskRunner({})->PostTask(
+        FROM_HERE,
+        base::BindOnce(&api::UtilityProcessWrapper::ShutdownStdioReader,
+                       utility_process_wrapper_,
+                       api::UtilityProcessWrapper::IOHandle::STDERR));
+  }
+
+  base::WeakPtr<api::UtilityProcessWrapper> utility_process_wrapper_;
+};
+
+namespace api {
+
+gin::WrapperInfo UtilityProcessWrapper::kWrapperInfo = {
+    gin::kEmbedderNativeGin};
+
+UtilityProcessWrapper::UtilityProcessWrapper(
+    node::mojom::NodeServiceParamsPtr params,
+    std::u16string display_name,
+    std::map<IOHandle, IOType> stdio,
+    base::EnvironmentMap env_map,
+    base::FilePath current_working_directory,
+    bool use_plugin_helper) {
+#if BUILDFLAG(IS_WIN)
+  base::win::ScopedHandle stdout_write(nullptr);
+  base::win::ScopedHandle stderr_write(nullptr);
+#elif BUILDFLAG(IS_POSIX)
+  base::FileHandleMappingVector fds_to_remap;
+#endif
+  for (const auto& [io_handle, io_type] : stdio) {
+    if (io_type == IOType::PIPE) {
+#if BUILDFLAG(IS_WIN)
+      HANDLE read = nullptr;
+      HANDLE write = nullptr;
+      if (!::CreatePipe(&read, &write, nullptr, 0)) {
+        PLOG(ERROR) << "pipe creation failed";
+        return;
+      }
+      if (io_handle == IOHandle::STDOUT) {
+        stdout_write.Set(write);
+        stdout_reader_ = std::make_unique<StdoutPipeReader>(
+            weak_factory_.GetWeakPtr(), read);
+        if (!stdout_reader_->Start()) {
+          PLOG(ERROR) << "stdout output watcher thread failed to start.";
+          PipeIOBase::Shutdown(std::move(stdout_reader_));
+          CloseHandle(write);
+          return;
+        }
+      } else if (io_handle == IOHandle::STDERR) {
+        stderr_write.Set(write);
+        stderr_reader_ = std::make_unique<StderrPipeReader>(
+            weak_factory_.GetWeakPtr(), read);
+        if (!stderr_reader_->Start()) {
+          PLOG(ERROR) << "stderr output watcher thread failed to start.";
+          PipeIOBase::Shutdown(std::move(stderr_reader_));
+          CloseHandle(write);
+          return;
+        }
+      }
+#elif BUILDFLAG(IS_POSIX)
+      int pipe_fd[2];
+      if (HANDLE_EINTR(pipe(pipe_fd)) < 0) {
+        PLOG(ERROR) << "pipe creation failed";
+        return;
+      }
+      if (io_handle == IOHandle::STDOUT) {
+        fds_to_remap.push_back(std::make_pair(pipe_fd[1], STDOUT_FILENO));
+        stdout_reader_ = std::make_unique<StdoutPipeReader>(
+            weak_factory_.GetWeakPtr(), pipe_fd[0]);
+        if (!stdout_reader_->Start()) {
+          PLOG(ERROR) << "stdout output watcher thread failed to start.";
+          PipeIOBase::Shutdown(std::move(stdout_reader_));
+          IGNORE_EINTR(close(pipe_fd[1]));
```

Comment:
Should the item in `fds_to_remap` be removed too?

Reply:
They should get removed after the return statement since the vector goes out of scope.

Reply:
Due to code chunk concern https://github.com/electron/electron/pull/34980#discussion_r980362923, I have reimplemented the reader using Node.js utilities https://github.com/electron/electron/pull/34980/commits/6a3c11c38ce51cdca155a3675cdded084ee03cf4. Thoughts on this implementation ?